specs/interop/superchain_backend.md

```diff
@@ -0,0 +1,205 @@
+dependents = dict() # (initiating_chain_id, block_number) -> List[(executing_chain_id, block_hash)]
+dependencies = dict() # (executing_chain_id, block_hash) -> List[(initiating_chain_id, block_number)]
+
+class Block:
+    header: Header
+    transactions: List[Transaction]
+
+def parse_executing_messages(block: Block):
+    """
+    Extract all executing messages in this block
+    """
+    messages = []
+    for tx in block.transactions:
+        if is_executing_tx(tx):
+            msg_id, payload = parse_executing_tx_data(tx.data)
+            messages.append(msg_id, payload)
+    return messages
+
+def block_safety(chain_id: int, block_hash: Bytes32):
+    """
+    Return the safety level for a given block, Used by the message safety backend API
+    to determine the safety of the remote block specified in the message identifier.
+    """
+    if len(block_unverified_executing_messages[block_hash]) > 0:
+        return UNSAFE
+
+    block = blocks[block_hash]
+
+    ## All messages were validated but we also need to ensure all transitive
+    ## block dependencies have had their messages also validated
+    for initiating_chain_id, block_number in dependencies[(chain_id, block_hash)]:
+
+        # dependency block has not been seen. We should never reach here as
+        # there should be no remaining unverified executing messages if tue. 
+        if block.header.number > chain_block_head[initiating_chain_id].header.number:
+            return UNSAFE
+
+        # check that transitive blocks are not unsafe
+        dependency_block = blocks[chain_block_number_to_hash[(initiating_chain_id, block_number)]]
+        if block_safety(initiating_chain_id, dependency_block.header.hash) == UNSAFE:
+            return UNSAFE
+
+    return CROSS_UNSAFE
+
+def add_unsafe_block(chain_id: int, block: Bytes32):
+    """
+    Extend the graph with a new block for the specified chain.
+    - Called when receiving an unsafe block over p2p from the sequencer
+    - Called when the safe head progress and the prior unsafe blocks was
+      not propogated.
+
+    For simplicity of this pseudocode, we exepct the caller to supply
+    the expected height and leave out reorg handling when either a new
+    unsafe block is supplied at the same height or the safe head mismatches
+    with a previously added unsafe block.
+    """
+    require(block.header.number == blocks[chain_block_head[chain_id]].header.number + 1)
+    require(block.header.parent_hash == chain_block_head[chain_id])
+
+    blocks[block.header.hash] = block
+    chain_block_head[chain_id] = block.header.hash
+
+    executing_messages = parse_executing_messages(block)
+    block_unverified_executing_messages[block.header.hash] = executing_messages
+
+    # add edges based on the executing messages present
+    for msg_id, _ in executing_messages:
+        # initiating block linked with the executing block
+        dependents[(msg_id.origin, msg_id.block_number)].add((chain_id, block.header.hash))
```

Comment:
I think we want to replace `msg_id.origin` with `chain_id` - see https://specs.optimism.io/interop/messaging.html#message-identifier

Reply:
ah yea that's right